### PR TITLE
Report whether video content is optimized for streaming.

### PR DESCRIPTION
**Pull request by skidder**
It could be useful to know whether the video content supports progressive download (`moov` atom is before the `mdat` atom/box) for optimized web streaming. Though this could be loosely inferred from the `AVFormatContext` flags, the most accurate method is to just inspect the MP4 atoms.

## Resources
[Handbrake docs](https://handbrake.fr/docs/en/latest/advanced/web-optimised.html)
[Understanding the MPEG-4 moov Atom](https://sanjeev-pandey.medium.com/understanding-the-mpeg-4-moov-atom-pseudo-streaming-in-mp4-93935e1b9e9a)
[Progressive Playback](https://fabiensanglard.net/mobile_progressive_playback/index.php)

## Benchmarks
Memory profile is unchanged, with an insignificant increase in time.

### Benchmark before adding check for streamable mp4
```
Running tool: /home/discord/bin/go/bin/go test -benchmem -run=^$ -coverprofile=/tmp/vscode-gof1InVD/go-code-cover -bench ^BenchmarkIsStreamableWebMp4$ github.com/discord/lilliput

goos: linux
goarch: amd64
pkg: github.com/discord/lilliput
cpu: Intel(R) Xeon(R) CPU @ 3.10GHz
BenchmarkIsStreamableWebMp4-16    	     765	   1537259 ns/op	     173 B/op	       4 allocs/op
PASS
coverage: 3.5% of statements
ok  	github.com/discord/lilliput	1.377s
```

### Benchmark with check for streamable mp4
```
Running tool: /home/discord/bin/go/bin/go test -benchmem -run=^$ -coverprofile=/tmp/vscode-goiNHYaO/go-code-cover -bench ^BenchmarkIsStreamableWebMp4$ github.com/discord/lilliput

goos: linux
goarch: amd64
pkg: github.com/discord/lilliput
cpu: Intel(R) Xeon(R) CPU @ 3.10GHz
BenchmarkIsStreamableWebMp4-16    	     729	   1540653 ns/op	     173 B/op	       4 allocs/op
PASS
coverage: 3.7% of statements
ok  	github.com/discord/lilliput	1.338s
```

### CODE REVIEW

File: avcodec.go [hunk 1]
```diff
@@ -58,6 +63,51 @@ func (d *avCodecDecoder) Description() string {
 	return fmt
 }
 
+func (d *avCodecDecoder) IsStreamable() bool {
```
Comment (brian-armstrong-discord) [inline avcodec.go:66]: neat!

let's move this into avcodec.cpp, maybe called from avcodec_decoder_create? we should also stash this as a flag somewhere so we don't potentially recompute it

File: avcodec.go [hunk 2]
```diff
@@ -58,6 +63,51 @@ func (d *avCodecDecoder) Description() string {
...
+			break
+		}
+		err = binary.Read(reader, binary.BigEndian, &atomType)
+		if err != nil {
+			break
+		}
+		bytesRead += int64(atomHeaderSize)
+
+		// Check for 'moov' atom
+		if string(atomType[:]) == "moov" {
```
Comment (walruscow) [inline avcodec.go:87]: does go do anything weird when making a string from invalid utf-8? might be safer to compare to a 4 byte array `'m', 'o', 'o', 'v'`

File: avcodec.go (hunk 1 above)
Comment (skidder) [reply avcodec.go:68]: sounds good, this approach came together nicely!

File: avcodec.go (hunk 2 above)
Comment (skidder) [reply avcodec.go]: Oh, great point. I'll ensure the C++ performs a byte-comparison of the atom name to avoid any string-interpretation.

**Review by brian-armstrong-discord — APPROVED** [review]
lgtm, let's try it out

**Pull request merged**